### ISSUE TITLE
关于GLN phase2训练细节的提问

### ISSUE THREAD

**Issue by 1994225000kk** [ShiqiYu/OpenGait#109]
你好！我在训练GLN的phase2时遇到了输出种类和预测种类数量不匹配的问题，debug排查了一下问题出现在compact_bloack的 logits 这一行：
 # compact_bloack
        if not self.pretrain:
            bn_feature = self.encoder_bn(feature.view(n, -1))  
            bn_feature = bn_feature.view(*feature.shape).contiguous()  

            reduce_feature = self.reduce_dp(bn_feature)  
            reduce_feature = self.reduce_ac(reduce_feature)  
            reduce_feature = self.reduce_fc(reduce_feature.view(n, -1))  

            bn_reduce_feature = self.reduce_bn(reduce_feature)  
            logits = self.reduce_cls(bn_reduce_feature).unsqueeze(1) 
            reduce_feature = reduce_feature.unsqueeze(1).contiguous()
            bn_reduce_feature = bn_reduce_feature.unsqueeze(1).contiguous()

1. 首先self.reduce_cls(bn_reduce_feature)得到的是[n,c], 如果在第2个维度unsqueeze一个维度就变成了[n,1,c]，无法和进入softmax代码时格式要求[n,c,p]相匹配导致了种类不匹配的报错。我试着改成unsqueeze(2) 让种类c对应输入格式[n,c,p]就可以运行，但精度没有论文所述明显的提升，请问这里是代码写错了嘛？
2.  这段代码的最后两行实际运行没有用到，是灰色的，不知是编写的时候测试忘记注释还是另有其用呢？
3.  另外，看了之前的提问者21年的GLN相关回复，编写者说在phase2阶段的compact_bloack训练也无法达到论文中的精度，请问现在有新进展了嘛？

如有时间 感谢您百忙中解答我的疑惑。

**Comment by noahshen98** [comment]
hi thanks for your question.

1. Since opengait has been updating year by year, the code of gln remained at the very initial version. It might be because the initial opengait calculate softmax loss in shape of (n p c)

2. It is useless, but we left it for testing.

3. I have not been able to reproduce the gln phase2, and I haven't kept following this method since the first commit.

**Comment by 1994225000kk** [comment]
Thank you for your reply.
